cpp/src/io/parquet/experimental/page_index_filter.cu

```diff
@@ -0,0 +1,854 @@
+      page_row_counts_and_offsets_tasks;
+    page_row_counts_and_offsets_tasks.reserve(num_columns);
+
+    auto streams = cudf::detail::fork_streams(stream, num_columns);
+
+    std::for_each(thrust::counting_iterator<size_t>(0),
+                  thrust::counting_iterator(num_columns),
+                  [&](auto const col_idx) {
+                    page_row_counts_and_offsets_tasks.emplace_back(
+                      cudf::detail::host_worker_pool().submit_task([&, col_idx = col_idx] {
+                        auto const schema_idx = output_column_schemas[col_idx];
+                        return make_page_row_counts_and_offsets(
+                          per_file_metadata, row_group_indices, schema_idx, streams[col_idx]);
+                      }));
+                  });
+
+    // Collect results from all tasks
+    std::for_each(page_row_counts_and_offsets_tasks.begin(),
+                  page_row_counts_and_offsets_tasks.end(),
+                  [&](auto& task) {
+                    auto [counts, offsets, chunk_offsets] = std::move(task).get();
+                    page_row_counts.emplace_back(std::move(counts));
+                    page_row_offsets.emplace_back(std::move(offsets));
+                    col_chunk_page_offsets.emplace_back(std::move(chunk_offsets));
+                  });
+  }
+
+  CUDF_EXPECTS(page_row_offsets.back().back() == total_rows,
+               "Mismatch in total rows in input row mask and row groups",
+               std::invalid_argument);
+
+  // Return if all rows are required or all are invalid.
+  if (row_mask.null_count() == row_mask.size() or thrust::all_of(rmm::exec_policy(stream),
+                                                                 row_mask.begin<bool>(),
+                                                                 row_mask.end<bool>(),
+                                                                 cuda::std::identity{})) {
+    return all_required_data_pages(page_row_counts);
+  }
+
+  auto const mr = cudf::get_current_device_resource_ref();
+
+  // Vector to hold data page mask for each column
+  auto data_page_mask = std::vector<thrust::host_vector<bool>>();
+  data_page_mask.reserve(num_columns);
+
+  std::atomic<size_t> total_surviving_pages{0};
+
+  // Tasks to compute data page mask for each column
+  std::vector<std::future<thrust::host_vector<bool>>> data_page_mask_tasks;
+  data_page_mask_tasks.reserve(num_columns);
+  auto streams = cudf::detail::fork_streams(stream, num_columns);
+
+  // For all columns, look up which pages contain at least one required row. i.e.
+  // !validity_it[row_idx] or is_row_required[row_idx] satisfies, and add its byte range to the
+  // output list of byte ranges for the column.
+  std::for_each(
+    thrust::counting_iterator<size_t>(0),
+    thrust::counting_iterator(num_columns),
+    [&](auto const col_idx) {
+      data_page_mask_tasks.emplace_back(
+        cudf::detail::host_worker_pool().submit_task([&, col_idx = col_idx] {
```

Comment:
Was seeing a segfault in the new gtests due to `col_idx` being invalid in some threads. Explicitly capturing it solves that bug. Not sure why we were seeing it in the first place.